vm/src/builtins.rs

```diff
@@ -559,11 +559,22 @@ fn builtin_pow(vm: &VirtualMachine, args: PyFuncArgs) -> PyResult {
     }
 }
 
+// Idea: Should we have a 'default' attribute, so we don't have to use OptionalArg's in this case
+//#[derive(Debug, FromArgs)]
+//pub struct PrintOptions {
+//    #[keyword] #[default(None)] sep: Option<PyStringRef>,
+//    #[keyword] #[default(None)] end: Option<PyStringRef>,
+//    #[keyword] #[default(flush)] flush: bool,
+//}
+
 #[derive(Debug, FromArgs)]
 pub struct PrintOptions {
-    sep: Option<PyStringRef>,
-    end: Option<PyStringRef>,
-    flush: bool,
+    #[keyword]
+    sep: OptionalArg<Option<PyStringRef>>,
```

Comment:
I would prefer to avoid using `OptionalArg` in these:

1. `OptionalArg<Option<...>>` feels very weird
2. We can derive whether an arg is optional or not based on the presence of a future `default` attribute.

Reply:
Agreed, OptionalArg<Option<...>> is awkward, but wouldn't be needed when I implement the default attribute argument. I'm split about having OptionalArg in these structs, I think in some cases we need to differentiate between a missing argument and passing None.

Reply:
I think we should handle this entirely with attributes rather than reusing `OptionalArg`. In the following, `None` refers to a Python `None` and `Option::None` refers to Rust's none.

Thing is required, and cannot be `None`.

```rust
#[derive(FromArgs)]
struct MyArgs {
    #[pyarg(positional)]
    thing: ThingRef,
}
```

Thing is required, but can be `None`:

```rust
#[derive(FromArgs)]
struct MyArgs {
    #[pyarg(positional)]
    thing: Option<ThingRef>,
}
```

Thing is not required, can be `None`, and will use the `Default` impl for `ThingRef` if not passed, which is `Option::None`.

```rust
#[derive(FromArgs)]
struct MyArgs {
    #[pyarg(positional, default)]
    thing: Option<ThingRef>,
}
```

Thing is not required, cannot be none, and will use the result of <method that returns ThingRef> if the argument is not present.

```rust
#[derive(FromArgs)]
struct MyArgs {
    #[pyarg(positional, default = "<method that returns ThingRef>")]
    thing: ThingRef,
}
```

Reply:
I still think in some situations it's not sufficient to just have a default, some functions need to detect if its argument was provided. For example, if the base argument is provided to int.__new__ then the first argument can then only be a string, otherwise, it can be other types like int or float.

CPython implements int.__new__  by using NULL as the default.  https://github.com/python/cpython/blob/master/Objects/longobject.c#L4950

I have added an 'optional 'argument to pyarg, to indicate when we are want to convert into OptionalArg. e.g.
```rust
#[derive(FromArgs)]
struct IntOptions {
    #[pyarg(positional, optional = true)]
    val_options: OptionalArg<PyObjectRef>,
    #[pyarg(positional_keyword, optional = true)]
    base: OptionalArg<u32>,
}
```

Reply:
> For example, if the base argument is provided to int.new then the first argument can then only be a string, otherwise, it can be other types like int or float.

In cases like these, the "function overloading" trick might work, e.g. range does:

```
fn range_new(vm: &VirtualMachine, args: PyFuncArgs) -> PyResult {
    let range = if args.args.len() <= 2 {
        let (cls, stop) = args.bind(vm)?;
        PyRangeRef::new(cls, stop, vm)
    } else {
        let (cls, start, stop, step) = args.bind(vm)?;
        PyRangeRef::new_from(cls, start, stop, step, vm)
    }?;

    Ok(range.into_object())
}
```

Basically there can be two different `int_new` methods (that accept different types) decided upon based on arity.